streams/src/main/java/org/apache/kafka/streams/processor/internals/DefaultStateUpdater.java

```diff
@@ -0,0 +1,373 @@
+            this.task = task;
+            this.action = action;
+        }
+    }
+
+    private final Time time;
+    private final ChangelogReader changelogReader;
+    private final java.util.function.Consumer<Set<TopicPartition>> offsetResetter;
+    private final Queue<TaskAndAction> tasksAndActions = new LinkedList<>();
+    private final Lock tasksAndActionsLock = new ReentrantLock();
+    private final Condition tasksAndActionsCondition = tasksAndActionsLock.newCondition();
+    private final Queue<StreamTask> restoredActiveTasks = new LinkedList<>();
+    private final Lock restoredActiveTasksLock = new ReentrantLock();
+    private final Condition restoredActiveTasksCondition = restoredActiveTasksLock.newCondition();
+    private final BlockingQueue<ExceptionAndTasks> failedTasks = new LinkedBlockingQueue<>();
+
+    private StateUpdaterThread stateUpdaterThread = null;
+
+    public DefaultStateUpdater(final ChangelogReader changelogReader,
+                               final java.util.function.Consumer<Set<TopicPartition>> offsetResetter,
+                               final Time time) {
+        this.changelogReader = changelogReader;
+        this.offsetResetter = offsetResetter;
+        this.time = time;
+    }
+
+    @Override
+    public void add(final Task task) {
+        if (stateUpdaterThread == null) {
+            stateUpdaterThread = new StateUpdaterThread("state-updater", changelogReader, offsetResetter);
+            stateUpdaterThread.start();
+        }
+
+        verifyStateFor(task);
+
+        tasksAndActionsLock.lock();
+        try {
+            tasksAndActions.add(new TaskAndAction(task, Action.ADD));
+            tasksAndActionsCondition.signalAll();
+        } finally {
+            tasksAndActionsLock.unlock();
+        }
+    }
+
+    private void verifyStateFor(final Task task) {
+        if (task.isActive() && task.state() != State.RESTORING) {
+            throw new IllegalStateException("Active task " + task.id() + " is not in state RESTORING. " + BUG_ERROR_MESSAGE);
+        }
+    }
+
+    @Override
+    public void remove(final Task task) {
+    }
+
+    @Override
+    public Set<StreamTask> getRestoredActiveTasks(final Duration timeout) {
+        final long timeoutMs = timeout.toMillis();
+        final long startTime = time.milliseconds();
+        final long deadline = startTime + timeoutMs;
+        long now = startTime;
+        final Set<StreamTask> result = new HashSet<>();
+        try {
+            while (now <= deadline && result.isEmpty()) {
+                restoredActiveTasksLock.lock();
+                try {
+                    while (restoredActiveTasks.isEmpty() && now <= deadline) {
+                        final boolean elapsed = restoredActiveTasksCondition.await(deadline - now, TimeUnit.MILLISECONDS);
+                        now = time.milliseconds();
+                    }
+                    while (!restoredActiveTasks.isEmpty()) {
```

Comment:
Can we just `result.addAll` and then `restoredActiveTasks.clear()` with just a `if` condition?

Reply:
Good point! We even do not need the `if` condition.